### PR TITLE
feat: ConfigService for centralizing configuration and simplifying BaseService

### PR DESCRIPTION
**Pull request by tbarlow12**
## What did you implement:

`ConfigService` and its usage. This removes the configuration logic from `BaseService` and keeps it all in one place. Also adds some constants for default configuration.

In a future PR, the `ConfigService` will cache its configuration so that the service is only configured one time in the initialization of the first `BaseService` subclass. It will store the data in the `serverless.variables` object and checks to see if that value is populated on future runs.

Also added some constants for default configuration values.

## How did you implement it:

Removed the configuration logic from the `BaseService` and placed inside the `ConfigService`, which is initialized in the constructor of the `BaseService`

## How can we verify it:

- Unit tests
- Run an end-to-end deployment of a service

## Todos:

_**Note: Run `npm run test:ci` to run all validation checks on proposed changes**_

- [x] Write tests and confirm existing functionality is not broken.  
       **Validate via `npm test`**
- [x] Write documentation
- [x] Ensure there are no lint errors.  
       **Validate via `npm run lint`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

File: src/services/configService.test.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      const actualResourceGroupName = service.getResourceGroupName();
+      const expectedRegion = AzureNamingService.createShortAzureRegionName(service.getRegion());
+      const expectedStage = AzureNamingService.createShortStageName(service.getStage());
+      const expectedResourceGroupName = `sls-${expectedRegion}-${expectedStage}-${serverless.service["service"]}-rg`;
+      expect(actualResourceGroupName).toEqual(expectedResourceGroupName);
+    });
+  });
+  
+  describe("Service Principal Configuration", () => {
+
```
Comment (wbreza) [inline src/services/configService.test.ts]: Nit: empty line

File: src/services/configService.test.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    config.service = serviceName;
+
+    delete config.provider.resourceGroup;
+    delete config.provider.region;
+    delete config.provider.stage;
+    delete config.provider.prefix;
+  });
+
+  describe("Configurable Variables", () => {
+
```
Comment (wbreza) [inline src/services/configService.test.ts]: nit: Empty line

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (mydiemho) [inline .vscode/launch.json]: why is this file added?

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
```
Comment (mydiemho) [inline package.json]: this file should not be modified by you

File: src/shared/constants.ts [hunk 5]
```diff
@@ -21,4 +21,12 @@ export const constants = {
   queueName: "queueName",
   xAzureSettings: "x-azure-settings",
   entryPoint: "entryPoint",
-}
\ No newline at end of file
+  variableKeys: {
```
Comment (mydiemho) [inline src/shared/constants.ts]: why `variableKeys`?

File: .vscode/launch.json (hunk 3 above)
Comment (tbarlow12) [reply .vscode/launch.json]: This allows for debugging without the vs code extension, which has been pretty spotty for me lately

File: src/shared/constants.ts (hunk 5 above)
Comment (tbarlow12) [reply src/shared/constants.ts]: Keys for the `serverless.variables` object. We had just been using strings

File: package.json (hunk 4 above)
Comment (tbarlow12) [reply package.json]: We want to start targeting `1.0.1-{pre-release}` correct?
